### PR TITLE
ci: Add testing and publishing of Docker image to Docker Hub

### LINKED ISSUE #9: Add distribution to Docker Hub with Docker's build-push-action GitHub Action
@tacaswell and I had discussed adding CI to test the build and distribute the image to Docker Hub using Docker's [`build-push-action`](https://github.com/docker/build-push-action) GitHub Action. An [example of how this is used can be seen in `pyhf`](https://github.com/scikit-hep/pyhf/blob/71d0c148a295e42f86300b7fe317aea56187ee34/.github/workflows/publish-docker.yml) which results in the [`pyhf/pyhf` image on Docker Hub](https://hub.docker.com/r/pyhf/pyhf/tags).

### PR DESCRIPTION
**Pull request by matthewfeickert**
Resolves #9 

Use Docker's [`build-push-action`](https://github.com/docker/build-push-action) GitHub Action to test the image build using the on `push` and `pull_request` events and also on a nightly CRON job that runs at 0:01 UTC. Additionally, add CI to build the Docker image and then publish it to Docker Hub under the tag 'latest'. If this is due to a tag being pushed, then also tag the image with the release tag and tag with 'latest-stable'. **edit:** Also now run weekly deployment of `latest` to Docker Hub as a cron job given the suggestions of @thomasjpfan. It can also be run _on demand_ by using the new [GitHub Actions `workflow_dispatch` system](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/).

This requires the addition of a Docker Hub username and password as GitHub secrets with the following names `DOCKER_USERNAME` and `DOCKER_PASSWORD` to work, which I can't do.

The CI can be seen running on my fork here: https://github.com/matthewfeickert/mpl-docker/actions [GitHub Actions Status: CI](https://github.com/matthewfeickert/mpl-docker/actions)

For an example of what the results of these CI patterns would give you in terms of tags c.f. [the `pyhf` Docker Hub tags](https://hub.docker.com/r/pyhf/pyhf/tags).

### CODE REVIEW

**Comment by matthewfeickert** [comment]
@tacaswell @dopplershift Once the Docker Hub account has been setup and the secrets are added to GitHub I think this is ready for review.

**Comment by matthewfeickert** [comment]
Thanks to @dopplershift for adding the secrets! :raised_hands:

File: README.md
Comment (matthewfeickert) [inline README.md]: @QuLogic These obviously don't render yet, as there isn't anything there until this PR gets merged to `master`. However, these patterns do work. I can show examples for some of my other projects if that is useful.

**Review by QuLogic — APPROVED** [review]
Seems good; probably easiest to just see it in practice.

**Comment by matthewfeickert** [comment]
Thanks for the review @QuLogic. :+1: Please let me know if there's anything else the matplotlib dev team would like to see before merge.

**Review by mrakitin — APPROVED** [review]
Looks great to me!

**Comment by matthewfeickert** [comment]
> Looks great to me!

Great! If there are no objections, if someone who has merge permissions can merge this in I have a few follow up PRs that are already queued up on other branches that I can open up once this goes in.

**Comment by mrakitin** [comment]
@tacaswell, do you have any concerns about it? If not, I'll merge the PR.

**Comment by matthewfeickert** [comment]
> @tacaswell, do you have any concerns about it? If not, I'll merge the PR.

@tacaswell @QuLogic just a ping on this for any further thoughts.

**Comment by QuLogic** [comment]
I don't have any merge rights here.

**Comment by matthewfeickert** [comment]
> I don't have any merge rights here.

Ah, sorry for the noise then.

**Comment by tacaswell** [comment]
Hmm, @QuLogic _should_ have merge rights here...

**Comment by tacaswell** [comment]
I merged it and fixed the ACL so it matches permissions on the main Maptlotlib repo.

**Pull request merged**
